Tests/AWSLambdaRuntimeCoreTests/LambdaMockClient.swift

```diff
@@ -0,0 +1,294 @@
+            case .waitingForNextEvent:
+                return .fail(.cannotCallNextEndpointWhenAlreadyWaitingForEvent)
+            case .handlerIsProcessing:
+                return .fail(.cannotCallNextEndpointWhenAlreadyProcessingAnEvent)
+            }
+        }
+
+        mutating func invoke(_ event: Event) -> InvokeAction {
+            switch self.state {
+            case .initialState, .handlerIsProcessing:
+                // next() hasn't been called yet. Add to the event queue.
+                self.eventQueue.append(event)
+                return .wait
+            case .waitingForNextEvent(let eventArrivedHandler):
+                // The runtime is already waiting for an event
+                self.state = .handlerIsProcessing(
+                    accumulatedResponse: [],
+                    eventProcessedHandler: event.eventProcessedHandler
+                )
+                return .readyToProcess(eventArrivedHandler)
+            }
+        }
+
+        mutating func writeResult(buffer: ByteBuffer) -> ResultAction {
+            switch self.state {
+            case .handlerIsProcessing(var accumulatedResponse, let eventProcessedHandler):
+                accumulatedResponse.append(buffer)
+                self.state = .handlerIsProcessing(
+                    accumulatedResponse: accumulatedResponse,
+                    eventProcessedHandler: eventProcessedHandler
+                )
+                return .readyForMore
+            case .initialState, .waitingForNextEvent:
+                return .fail(.cannotReportResultWhenNoEventHasBeenProcessed)
+            }
+        }
+
+        mutating func finish() throws {
+            switch self.state {
+            case .handlerIsProcessing(let accumulatedResponse, let eventProcessedHandler):
+                let finalResult: ByteBuffer = accumulatedResponse.reduce(ByteBuffer()) { (accumulated, current) in
+                    var accumulated = accumulated
+                    accumulated.writeBytes(current.readableBytesView)
+                    return accumulated
+                }
+
+                eventProcessedHandler.resume(returning: finalResult)
+                // reset back to the initial state
+                self.state = .initialState
+            case .initialState, .waitingForNextEvent:
+                throw LambdaError.cannotReportResultWhenNoEventHasBeenProcessed
+            }
+        }
+
+        mutating func cancelNext() -> CancelNextAction {
+            switch self.state {
+            case .initialState, .handlerIsProcessing:
+                return .none
+            case .waitingForNextEvent(let eventArrivedHandler):
+                self.state = .initialState
+                return .cancelContinuation(eventArrivedHandler)
+            }
+        }
+
+        mutating func failProcessing() -> FailProcessingAction {
+            switch self.state {
+            case .initialState, .waitingForNextEvent:
+                return .none
```

Comment:
we should crash here, because this would mean, that we send a failProcessing at the wrong time.